js/blocks/components/fetch-input.js

```diff
@@ -27,31 +27,57 @@ const { addQueryArgs } = wp.url;
 const stopEventPropagation = ( event ) => event.stopPropagation();
 
 class FetchInput extends Component {
-
 	/**
 	 * Constructs the component class.
 	 */
 	constructor( { autocompleteRef } ) {
 		super( ...arguments );
 
-		this.onBlur            = this.onBlur.bind( this );
-		this.onFocus           = this.onFocus.bind( this );
-		this.onChange          = this.onChange.bind( this );
-		this.onKeyDown         = this.onKeyDown.bind( this );
-		this.autocompleteRef   = autocompleteRef || createRef();
-		this.inputRef          = createRef();
+		this.onBlur = this.onBlur.bind( this );
+		this.onFocus = this.onFocus.bind( this );
+		this.onChange = this.onChange.bind( this );
+		this.onKeyDown = this.onKeyDown.bind( this );
+		this.autocompleteRef = autocompleteRef || createRef();
+		this.inputRef = createRef();
 		this.updateSuggestions = this.updateSuggestions.bind( this );
-		this.setInputValidity  = this.setInputValidity.bind( this );
+		this.setInputValidity = this.setInputValidity.bind( this );
 
 		this.suggestionNodes = [];
 
 		this.state = {
+			loading: false,
 			results: [],
 			showSuggestions: false,
 			selectedSuggestion: null,
 		};
 	}
 
+	/**
+	 * Conditionally sets the validity of the <input>.
+	 *
+	 * Runs when the component updates, like with a change of state.
+	 *
+	 * @param {Object} prevProps The previous props.
+	 * @param {Object} prevState The previous state.
+	 */
+	componentDidUpdate( prevProps, prevState ) {
+		const { loading, results } = this.state;
+		const { prevLoading, prevResults } = prevState;
+
+		// Exit if the relevant state values didn't update.
+		if ( loading === prevLoading && results === prevResults ) {
+			return;
+		}
+
+		if ( ! loading ) {
+			if ( results.length ) {
+				this.setInputValidity( true );
+			} else {
+				this.setInputValidity( false );
+			}
+		}
+	}
```

Comment:
This method is intended to replace the `setInputValidity()` calls in the `render()` method. I shouldn't have added those to that before 😄 

This simplifies how the validity message appears.

Now, it mainly appears when entering something in the `<input>` that doesn't match any post (or taxonomy, or whatever control this is for):

<img width="883" alt="input-does-not-match-anything" src="https://user-images.githubusercontent.com/4063887/64915361-42c02e00-d72a-11e9-8479-30ddbc702f80.png">

It won't show this validation message anymore if there simply isn't anything in the `<input>`:

<img width="836" alt="validation-message" src="https://user-images.githubusercontent.com/4063887/64915365-671c0a80-d72a-11e9-92cb-446ab5985a22.png">

Still, maybe this could use more thought.